// @ngx-share/button/bundles/ngx-share-button.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ngx-share/core')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@ngx-share/core'], factory) :
	(factory((global['ngx-share'] = global['ngx-share'] || {}, global['ngx-share'].button = {}),global.ng.core,global.core$1));
}(this, (function (exports,core,core$1) { 'use strict';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var ShareButtonComponent = (function () {
    /**
     * @param {?} share
     */
    function ShareButtonComponent(share) {
        this.share = share;
        /**
         * Show button icon
         */
        this.showIcon = true;
        /**
         * Show button text
         */
        this.showText = false;
        /**
         * Button share count
         */
        this.showCount = false;
        /**
         * Button size
         */
        this.size = this.share.size;
        /**
         * Share count event
         */
        this.count = new core.EventEmitter();
        /**
         * Share dialog opened event
         */
        this.opened = new core.EventEmitter();
        /**
         * Share dialog closed event
         */
        this.closed = new core.EventEmitter();
        /**
         * Set theme as button class
         */
        this.buttonClass = 'sb-button sb-' + this.share.theme;
    }
    Object.defineProperty(ShareButtonComponent.prototype, "createButton", {
        /**
         * @param {?} button
         * @return {?}
         */
        set: function (button) {
            this.shareCount = 0;
            this.button = button;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareButtonComponent.prototype, "setUrl", {
        /**
         * on set share URL
         * @param {?} newUrl
         * @return {?}
         */
        set: function (newUrl) {
            /** Reset share count when url changes */
            this.shareCount = 0;
            this.url = newUrl;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareButtonComponent.prototype, "setTheme", {
        /**
         * Button theme
         * @param {?} theme
         * @return {?}
         */
        set: function (theme) {
            this.buttonClass = 'sb-button sb-' + theme;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} count
     * @return {?}
     */
    ShareButtonComponent.prototype.onCount = function (count) {
        this.shareCount = count;
        this.count.emit(count);
    };
    return ShareButtonComponent;
}());
ShareButtonComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'share-button',
                template: "<button class=\"sb-wrapper\"\n        [shareButton]=\"button\"\n        [sbUrl]=\"url\"\n        [sbImage]=\"image\"\n        [sbTitle]=\"title\"\n        [sbDescription]=\"description\"\n        [sbTags]=\"tags\"\n        (sbOpened)=\"opened.emit($event)\"\n        (sbClosed)=\"closed.emit($event)\"\n        (sbCount)=\"onCount($event)\"\n        [class.sb-show-icon]=\"showIcon\"\n        [class.sb-show-text]=\"showText\"\n        [class.sb-show-count]=\"showCount && shareCount\"\n        [style.fontSize.px]=\"(1 + size/20) * 14\">\n\n  <div class=\"sb-inner\">\n\n    <!-- HIDE BUTTON'S ICON AND TEXT IF CUSTOM TEMPLATE IS USED -->\n\n    <div class=\"sb-content\">\n\n      <!-- BUTTON ICON -->\n      <div *ngIf=\"showIcon\" class=\"sb-icon\">\n        <i [class]=\"ref.prop.icon\" aria-hidden=\"true\"></i>\n      </div>\n\n      <!-- BUTTON TEXT -->\n      <div *ngIf=\"showText\" class=\"sb-text\">\n        {{ text || ref.prop.text }}\n      </div>\n\n    </div>\n\n    <!-- BUTTON COUNT -->\n    <div *ngIf=\"showCount && shareCount\" class=\"sb-count\">\n      <span>{{ shareCount | nFormatter }}</span>\n    </div>\n\n  </div>\n</button>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
ShareButtonComponent.ctorParameters = function () { return [
    { type: core$1.ShareButtons, },
]; };
ShareButtonComponent.propDecorators = {
    "createButton": [{ type: core.Input, args: ['button',] },],
    "setUrl": [{ type: core.Input, args: ['url',] },],
    "title": [{ type: core.Input },],
    "description": [{ type: core.Input },],
    "image": [{ type: core.Input },],
    "tags": [{ type: core.Input },],
    "showIcon": [{ type: core.Input },],
    "showText": [{ type: core.Input },],
    "showCount": [{ type: core.Input },],
    "text": [{ type: core.Input },],
    "size": [{ type: core.Input },],
    "setTheme": [{ type: core.Input, args: ['theme',] },],
    "count": [{ type: core.Output },],
    "opened": [{ type: core.Output },],
    "closed": [{ type: core.Output },],
    "buttonClass": [{ type: core.HostBinding, args: ['class',] },],
    "ref": [{ type: core.ViewChild, args: [core$1.ShareButtonDirective,] },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @param {?} config
 * @return {?}
 */
function ShareButtonsFactory(config) {
    return new core$1.ShareButtons(config);
}
var ShareButtonModule = (function () {
    function ShareButtonModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    ShareButtonModule.forRoot = function (config) {
        return {
            ngModule: ShareButtonModule,
            providers: [
                { provide: core$1.CONFIG, useValue: config },
                {
                    provide: core$1.ShareButtons,
                    useFactory: ShareButtonsFactory,
                    deps: [core$1.CONFIG]
                }
            ]
        };
    };
    return ShareButtonModule;
}());
ShareButtonModule.decorators = [
    { type: core.NgModule, args: [{
                declarations: [
                    ShareButtonComponent
                ],
                imports: [
                    core$1.ShareModule
                ],
                exports: [
                    core$1.ShareModule,
                    ShareButtonComponent
                ]
            },] },
];
/** @nocollapse */
ShareButtonModule.ctorParameters = function () { return []; };

exports.ShareButtonsFactory = ShareButtonsFactory;
exports.ShareButtonModule = ShareButtonModule;
exports.ShareButtonComponent = ShareButtonComponent;

Object.defineProperty(exports, '__esModule', { value: true });

})));
